### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by PrudnikovV**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+version: 2.1
+
+executors:
+  default:
+    working_directory: ~/repo
+    docker:
+      - image: cimg/ruby:3.1-browsers
```
Comment (kolyanjy) [inline .circleci/config.yml]: I think we don't need browsers in image

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      - defaults
+      - run_specs
+      
+workflows:
+  build:
+    jobs:
+      - lintering
+      - run_specs:
+          requires:
+            - lintering
```
Comment (kolyanjy) [inline .circleci/config.yml]: add empty line

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+/coverage/
+/doc/
+/pkg/
+/spec/reports/
+/tmp/
+
+# rspec failure tracking
+.rspec_status
+data.yml
+test_spec.yml
```
Comment (kolyanjy) [inline .gitignore]: what is that?

File: .overcommit.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+PreCommit:
+  RuboCop:
+    enabled: true
+    on_warn: fail # Treat all warnings as failures
+
+PreCommit:
+  Fasterer:
+    enabled: true
+    on_warn: fail # Treat all warnings as failures
+
```
Comment (kolyanjy) [inline .overcommit.yml:27]: Good thing 👍 
also you can take a look on this [one](https://github.com/evilmartians/lefthook)

File: lib/attempt.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  class Attempt
+    attr_reader :result, :attempt_code, :secret_code
+
+    def initialize(attempt_code, secret_code)
+      @attempt_code = attempt_code
+      @secret_code = secret_code
+      colculation
+    end
+
+    def colculation
```
Comment (kolyanjy) [inline lib/attempt.rb]: tipo: calculation

File: lib/attempt.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      @result = @attempt_code.select.with_index { |e, i| e == @secret_code[i] }.map { "+" }
+      @result += ["-"] * (@attempt_code.uniq.map { |e| min_quentyty(e) }.sum - result.count)
```
Comment (kolyanjy) [inline lib/attempt.rb]: move '+' and '-' to constants

File: lib/attempt.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    end
+
+    private
+
+    def min_quentyty(elm)
```
Comment (kolyanjy) [inline lib/attempt.rb]: tipo: min_quantity

File: lib/codebreaker.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+require_relative("user")
+require_relative("difficulty")
+require_relative("statistic")
+require_relative("validation")
+require_relative "codebreaker/version"
+require "yaml"
+require "pry"
+
+module Codebreaker
+  class Error < StandardError; end
```
Comment (kolyanjy) [inline lib/codebreaker.rb]: move to separate file and use more correct naming

File: lib/difficulty.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Codebreaker
+  # class for Hint
+  class Difficulty
+    attr_reader :attempts, :hints, :name
+
```
Comment (kolyanjy) [inline lib/difficulty.rb:7]: I think here must be suite with hints and attempts quantity

File: lib/winner.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+    def initialize(name, difficulty, game)
+      @name = name
+      @difficulty = difficulty.name
+      @attempts = difficulty.attempts
+      @hints = difficulty.hints
+      @attempts_used = game.attempts.size
+      @hints_used = game.hints.size
+    end
+  end
```
Comment (kolyanjy) [inline lib/winner.rb]: i think this class redundant. we can just save user which win game

File: lib/game.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+      @remaining_attempts -= 1
+    end
+
+    def hint(hint)
+      @hints << hint
+      @remaining_hints -= 1
+    end
+
+    def new_secret_code
+      Array.new(4) { rand(1..6) }
```
Comment (kolyanjy) [inline lib/game.rb]: Pls move range to constant

File: spec/game_spec.rb [hunk 12]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+RSpec.describe Codebreaker::Game do
+  let(:difficulty) { Codebreaker::Difficulty.new("Easy", 15, 2) }
```
Comment (kolyanjy) [inline spec/game_spec.rb]: you just need to provide name of difficulty. quantity of attempts and hints must be resolved in difficulty class

File: lib/difficulty.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+    def initialize(name, attempts, hints)
```
Comment (kolyanjy) [inline lib/difficulty.rb]: just name

**Review by kolyanjy — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 3 above)
Comment (PrudnikovV) [reply .gitignore]: This was by default, after "bundle gem codebraker"

File: .circleci/config.yml [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
```
Comment (kolyanjy) [inline .circleci/config.yml:62]: pls add rule with last empty line to rubocop for avoid it
image

File: lib/codebreaker.rb (hunk 8 above)
Comment (kolyanjy) [reply lib/codebreaker.rb]: Still not resolved.

File: lib/difficulty.rb [hunk 15]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module Codebreaker
+  # class for Hint
+  class Difficulty
+    attr_reader :attempts, :hints, :name
+
+    STANDART_DIFFICULTIES =
```
Comment (kolyanjy) [inline lib/difficulty.rb:8]: 👍

**Review by kolyanjy — APPROVED** [review]
Fix comments and merge

**Pull request merged**